_posts/2025-01-13-mcp-server.adoc

```diff
@@ -0,0 +1,263 @@
+
+    // Get detailed forecast using dynamically provided URL
+    @GET
+    @Path("/")
+    Forecast getForecast(@Url String url);
+}
+----
+
+To handle the API responses, we'll define some data classes. Note that we're only including the fields we need, as the complete API response contains much more data:
+
+[source,java]
+----
+static record Period(
+    String name,
+    int temperature,
+    String temperatureUnit,
+    String windSpeed,
+    String windDirection,
+    String detailedForecast) {
+}
+
+static record ForecastProperties(
+        List<Period> periods) {
+}
+
+static record Forecast(
+        ForecastProperties properties) {
+}
+----
+
+Since the Weather API uses redirects, add this to your `application.properties`:
+
+[source,properties]
+----
+quarkus.rest-client.follow-redirects=true
+----
+
+=== Formatting Helpers
+
+We'll use Qute templates to format the weather data:
+
+[source,java]
+----
+String formatForecast(Forecast forecast) {
+    return forecast.properties().periods().stream().map(period -> {
+            // Template for each forecast period
+            return Qute.fmt(
+                """
+                        Temperature: {p.temperature}°{p.temperatureUnit}
+                        Wind: {p.windSpeed} {p.windDirection}
+                        Forecast: {p.detailedForecast}
+                        """,
+                Map.of("p", period)).toString();
+        }).collect(Collectors.joining("\n---\n"));
+    }
+----
+
+=== Implementing MCP Tools
+
+Now let's implement the actual MCP tools. The `@Tool` annotation from `io.quarkiverse.mcp.server` marks methods as available tools, while `@ToolArg` describes the parameters:
+
+[source,java]
+----
+@Tool(description = "Get weather alerts for a US state.")
+String getAlerts(@ToolArg(description = "Two-letter US state code (e.g. CA, NY)") String state) {
+    return formatAlerts(weatherClient.getAlerts(state));
+}
+
+@Tool(description = "Get weather forecast for a location.")
+String getForecast(
+    @ToolArg(description = "Latitude of the location") double latitude,
+    @ToolArg(description = "Longitude of the location") double longitude) {
+    
+    // First get the point metadata which contains the forecast URL
+    var points = weatherClient.getPoints(latitude, longitude);
+    // Extract the forecast URL using Qute template
+    var url = Qute.fmt("{p.properties.forecast}", Map.of("p", points));
+    // Get and format the forecast
+    return formatForecast(weatherClient.getForecast(url));
+}
+----
+
+[NOTE]
+====
+The forecast API requires a two-step process where we first get point metadata and then use a URL from that response to fetch the actual forecast.
+====
+
+== Running the Server
+
+To simplify deployment and development, we'll package the server as an uber-jar and enable file logging (since stdio is reserved for the MCP protocol):
+
+[source,properties]
+----
+quarkus.package.uber-jar=true
```

Comment:
Why uber jar?

Reply:
To run and distribute it easily with maven/java too. Not just jbang. But sure can put comment.